docs/proposals/dehydration.md

```diff
@@ -0,0 +1,147 @@
+1. Controllers need to be aware of a cluster being dehydrated and should skip reconciling them.
+
+It is likely that for the first two steps, a dehydration phase indicator needs to be added to the `Cluster` status, since the third step will not allow to interact with objects on the user cluster anymore and the previous steps need to be skipped at that point. It could look something like:
+
+```yaml
+status:
+  dehydration:
+    phase: MachinesScaledDown
+---
+status:
+  dehydration:
+    phase: Active
+```
+
+The condition that other controllers would react to could look like this:
+
+```yaml
+status:
+  conditions:
+    ClusterDehydrated:
+      kubermaticVersion: v2.22.0-23-gbc5f9c5f6
+      lastHeartbeatTime: "2023-03-08T09:28:12Z"
+      status: "True"
+    [...]
+```
+
+To reverse the dehydration (to "hydrate" a cluster again), the changes to control plane components need to be undone. This can be achieved by removing the condition that prevents controllers from reconciling the dehydrated cluster, which should cause the control plane components to be scaled up again. Once the cluster is up again, the `(de-)hydration-controller` needs to restore previous state of `MachineDeployments` by reading the annotation and scaling them as appropriate.
+
+#### Cluster Status before Dehydration
+
+Of particular concern is how to keep users and admins informed of what a user cluster was like before it got dehydrated. This includes key information like how many `PersistentVolumes`, `LoadBalancers` and `Machines` existed in the cluster, since shutting down the control plane means there is no way to enumerate them before hydrating again.
+
+Because of that, adding a status field to `Clusters` might be beneficial to include some basic information. This will not give the full details, but rather an "executive summary" of what the cluster looks like, what is still incurring costs and what it means to hydrate the cluster again. The status field could look something like this:
+
+```yaml
+apiVersion: kubermatic.k8c.io/v1
+kind: Cluster
+metadata:
+  name: xyz1234
+spec:
+  humanReadableName: dehydration-cluster
+  dehydrated: true
+  [...]
+status:
+  dehydration:
+    phase: Active
+    remaining:
+      volumes: 3
+      loadBalancers: 2
+    scaledDownMachines: 5
+```
+
+This information should be shown in the UI in some way, to highlight that some resources associated with this cluster likely still exist (which is what the feature is supposed to ensure). `status.dehydration.scaledDownMachines` can be used in the dialog to re-hydrate a cluster again, letting the user know that x machines will be created when hydrating the user cluster.
+
+### Limitations
+
+* "Bring Your Own" clusters cannot be dehydrated because KKP does not control the nodes and cannot scale them down. Dehydrating such a cluster is therefore useless.
+* Dehydrated user clusters will not count towards resource quotas, since the nodes will be removed.
```

Comment:
How about storage? It can be also defined in `ResoureQuota`. If we want to keep disks around I guess we should only subtract cpu and memory.

Reply:
what counts towards the storage quota? OS disks for machines? Persistent volumes? And how does the quota get adjusted, if the user cluster goes "down" due to dehydration, will the quota stay as-is or will it subtract things it cannot find anymore?

Reply:
The resource quota docs say:

> Storage - the cumulated disk size of the nodes on all clusters.

So this seems to only refer to node disks, which will be deleted as part of the machine scale-down.

Reply:
That makes sense then, I thought that we want to keep node disks around for the time of de-hydration. But I guess that would be hard to implement and very provider specific.

Reply:
> what counts towards the storage quota? OS disks for machines? Persistent volumes? And how does the quota get adjusted, if the user cluster goes "down" due to dehydration, will the quota stay as-is or will it subtract things it cannot find anymore?

This is something to check. I believe the controller is watching Machine resources and acts upon changes there.

Reply:
> This is something to check. I believe the controller is watching Machine resources and acts upon changes there.

That would certainly match our documentation.